applets/wncklet/window-list.c

```diff
@@ -255,50 +257,71 @@ static GdkPixbuf *preview_window_thumbnail (WnckWindow *wnck_window, TasklistDat
 	width = gdk_window_get_width (window) * scale;
 	height = gdk_window_get_height (window) * scale;
 
-	/* Generate window screenshot for preview */
-	screenshot = gdk_pixbuf_get_from_window (window, 0, 0, width / scale, height / scale);
+	/* Get reference to GdkWindow surface */
+	cairo_t *win_cr = gdk_cairo_create (window);
+	cairo_surface_t *win_surface = cairo_get_target (win_cr);
+	/* Flush to ensure all writing to the image was done */
+	cairo_surface_flush (win_surface);
+	cairo_destroy (win_cr);
+
+	/* Create screenshot surface with the GdkWindow as its source */
+	screenshot = cairo_image_surface_create (CAIRO_FORMAT_ARGB32, width, height);
+	cairo_surface_set_device_scale (screenshot, scale, scale);
+	cr = cairo_create (screenshot);
+	cairo_set_source_surface (cr, win_surface, 0, 0);
+	cairo_paint (cr);
+
+	/* Mark the image dirty so Cairo clears its caches */
+	cairo_surface_mark_dirty (win_surface);
+
+	cairo_destroy (cr);
 	g_object_unref (window);
 
 	if (screenshot == NULL)
 		return NULL;
 
-	/* Determine whether the contents of the screenshot are empty */
-	if (gdk_pixbuf_get_byte_length (screenshot) == 0)
-	{
-		g_object_unref (screenshot);
-		return NULL;
-	}
-
 	/* Scale to configured size while maintaining aspect ratio */
 	if (width > height)
 	{
-		ratio = (double) height / (double) width;
-		width = MIN(width, tasklist->thumbnail_size);
-		height = width * ratio;
+		int max_size = MIN(width, tasklist->thumbnail_size);
+		ratio = (double) width / (double) max_size;
+		width = max_size;
+		height = (int) ((double) height / ratio);
 	}
 	else
 	{
-		ratio = (double) width / (double) height;
-		height = MIN(height, tasklist->thumbnail_size);
-		width = height * ratio;
+		int max_size = MIN(height, tasklist->thumbnail_size);
+		ratio = (double) height / (double) max_size;
+		height = max_size;
+		width = (int) ((double) width / ratio);
 	}
 
-	thumbnail = gdk_pixbuf_scale_simple (screenshot, width, height, GDK_INTERP_BILINEAR);
-	g_object_unref (screenshot);
+	thumbnail = cairo_surface_create_similar (screenshot,
+						  cairo_surface_get_content (screenshot),
+						  width, height);
+
+	cr = cairo_create (thumbnail);
+	cairo_scale (cr, scale/ratio, scale/ratio);
+	cairo_set_source_surface (cr, screenshot, 0, 0);
+	cairo_paint (cr);
+
+	cairo_destroy (cr);
+	cairo_surface_destroy (screenshot);
 
 	return thumbnail;
 }
 
 #define PREVIEW_PADDING 5
-static void preview_window_reposition (TasklistData *tasklist, GdkPixbuf *thumbnail)
+static void preview_window_reposition (TasklistData *tasklist, cairo_surface_t *thumbnail)
 {
 	GdkMonitor *monitor;
 	GdkRectangle monitor_geom;
 	int x_pos, y_pos;
-	int width, height;
+	int width, height, scale;
 
-	width = gdk_pixbuf_get_width (thumbnail);
-	height = gdk_pixbuf_get_height (thumbnail);
+	scale = gtk_widget_get_scale_factor (tasklist->preview);
```

Comment:
What happens if the scale of the thumbnail is different from its placement?

Reply:
not sure I understand the question. Scale is based on the dpi multiplier (1 for regular, 2 for HiDPI, etc). The width/height are used to resize the thumbnail correctly, but the positioning is done based on the mouse coordinates. Without the scale, in a HiDPI environment, the thumbnail will end up being a large widget with a small surface in its corner.

Reply:
In this example we have two screens: The main screen is not HiDPI so it has a 1:1 scale. The second screen is HiDPI and it has a 2:1 scale. The window thumbnails of the second screen will have twice the normal size if they are displayed on the main screen.

Reply:
If the scale has the same value for all screens, you could set its value within the function `applet_enter_notify_event`, and call the two functions (`preview_window_thumbnail` and `preview_window_reposition`) adding this new argument.

Reply:
```C
diff --git a/applets/wncklet/window-list.c b/applets/wncklet/window-list.c
index 4708d56c..a456bac2 100644
--- a/applets/wncklet/window-list.c
+++ b/applets/wncklet/window-list.c
@@ -237,7 +237,7 @@ static void applet_change_background(MatePanelApplet* applet, MatePanelAppletBac
 
 #ifdef HAVE_X11
 #ifdef HAVE_WINDOW_PREVIEWS
-static cairo_surface_t *preview_window_thumbnail (WnckWindow *wnck_window, TasklistData *tasklist)
+static cairo_surface_t *preview_window_thumbnail (WnckWindow *wnck_window, TasklistData *tasklist, int *thumbnail_width, int *thumbnail_height, int *thumbnail_scale)
 {
        GdkWindow *window;
        cairo_surface_t *screenshot;
@@ -306,20 +306,19 @@ static cairo_surface_t *preview_window_thumbnail (WnckWindow *wnck_window, Taskl
        cairo_destroy (cr);
        cairo_surface_destroy (screenshot);
 
+       *thumbnail_width = width;
+       *thumbnail_height = height;
+       *thumbnail_scale = scale;
+
        return thumbnail;
 }
 
 #define PREVIEW_PADDING 5
-static void preview_window_reposition (TasklistData *tasklist, cairo_surface_t *thumbnail)
+static void preview_window_reposition (TasklistData *tasklist, cairo_surface_t *thumbnail, int width, int height)
 {
        GdkMonitor *monitor;
        GdkRectangle monitor_geom;
        int x_pos, y_pos;
-       int width, height, scale;
-
-       scale = gtk_widget_get_scale_factor (tasklist->preview);
-       width = cairo_image_surface_get_width (thumbnail) / scale;
-       height = cairo_image_surface_get_height (thumbnail) / scale;
 
        /* Resize window to fit thumbnail */
        gtk_window_resize (GTK_WINDOW (tasklist->preview), width, height);
@@ -368,6 +367,9 @@ static gboolean applet_enter_notify_event (WnckTasklist *tl, GList *wnck_windows
        cairo_surface_t *thumbnail;
        WnckWindow *wnck_window = NULL;
        int n_windows;
+       int thumbnail_width;
+       int thumbnail_height;
+       int thumbnail_scale;
 
        if (tasklist->preview != NULL)
        {
@@ -395,7 +397,7 @@ static gboolean applet_enter_notify_event (WnckTasklist *tl, GList *wnck_windows
                                                  wnck_screen_get_active_workspace (wnck_screen_get_default ())))
                return FALSE;
 
-       thumbnail = preview_window_thumbnail (wnck_window, tasklist);
+       thumbnail = preview_window_thumbnail (wnck_window, tasklist, &thumbnail_width, &thumbnail_height, &thumbnail_scale);
 
        if (thumbnail == NULL)
                return FALSE;
@@ -406,7 +408,7 @@ static gboolean applet_enter_notify_event (WnckTasklist *tl, GList *wnck_windows
        gtk_widget_set_app_paintable (tasklist->preview, TRUE);
        gtk_window_set_resizable (GTK_WINDOW (tasklist->preview), TRUE);
 
-       preview_window_reposition (tasklist, thumbnail);
+       preview_window_reposition (tasklist, thumbnail, thumbnail_width/thumbnail_scale, thumbnail_height/thumbnail_scale);
 
        gtk_widget_show (tasklist->preview);
 
```

Reply:
It can be done in another commit.